src/celeritas/optical/action/detail/TrackInitAlgorithms.cu

```diff
@@ -53,18 +54,73 @@ size_type copy_if_vacant(TrackStatusRef<MemSpace::device> const& status,
 
     ScopedProfiling profile_this{"copy-if-vacant"};
 
+    using StreamT = CELER_DEVICE_API_SYMBOL(Stream_t);
+    Stream& s = device().stream(stream_id);
+    StreamT stream = s.get();
+
+    void* d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    // *** For testing with existing code for consistency
+    size_type* d_num_vacancies = nullptr;
+    // *** Allocate temporary storage
+    // *** For testing with existing code for consistency
+    d_num_vacancies
+        = static_cast<size_type*>(s.malloc_async(sizeof(size_type)));
+
+    // The first call just computes the number of additional bytes needed for
+    // the in-place selection. The nullptr value causes this instead of running
+    // the function.
     auto start = thrust::make_transform_iterator(
         thrust::make_counting_iterator<size_type>(0), TransformType{});
     auto result = device_pointer_cast(vacancies.data());
-    auto end = thrust::copy_if(thrust_execute_on(stream_id),
-                               start,
-                               start + vacancies.size(),
-                               device_pointer_cast(status.data()),
-                               result,
-                               IsVacant{});
+    auto flags = device_pointer_cast(status.data());
+    cub::DeviceSelect::FlaggedIf(d_temp_storage,
+                                 temp_storage_bytes,
+                                 start,
+                                 flags,
+                                 result,
+                                 d_num_vacancies,
+                                 vacancies.size(),
+                                 IsVacant{},
+                                 stream);
+
+    // Allocate temporary storage
+    d_temp_storage = s.malloc_async(temp_storage_bytes);
```

Comment:
Since `temp_storage_bytes` is a stack value, doesn't this introduce a synchronization even if the previous kernel and the following allocation are async?